### PR TITLE
#61 Interaction with the Github GraphQL API

### PR DESCRIPTION
**Pull request by Suban05**
## Changes

* Added class `Fbe::Graph`, which allows interacting with GitHub.
* Added a method `Fbe.github_graph`. It creates an instance of `Fbe::Graph` and prepares it for use. It works like `Fbe.octo`.
* The `resolved_conversations` and `total_commits` methods have been added to `Fbe.github_graph`.
* Added tests

### CODE REVIEW

**Comment by Suban05** [comment]
@Yegorov please review

File: lib/fbe/github/graphql/client.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# SOFTWARE.
+
+require 'graphql/client'
+
+# The GitHub GraphQL client
+class Fbe::GitHub::GraphQL::Client
+  def initialize(token:)
+    http = Fbe::GitHub::GraphQL::HTTP.new(token)
+    schema = GraphQL::Client.load_schema(http)
+    @client = GraphQL::Client.new(schema:, execute: http)
```
Comment (Yegorov) [inline lib/fbe/github/graphql/client.rb:32]: Looks as [redundant variable](https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html)
Maybe do as:
```ruby
@client = GraphQL::Client.new(schema: GraphQL::Client.load_schema(http), execute: http)
```

File: lib/fbe/github/graphql/client.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    @client.allow_dynamic_queries = true
+  end
+
+  def query(query_string)
+    query  = @client.parse(query_string)
+    result = @client.query(query)
```
Comment (Yegorov) [inline lib/fbe/github/graphql/client.rb:38]: The same [redundant variable](https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html):
```ruby
result = @client.query(@client.parse(query_string))
```

File: lib/fbe/github/graphql/http.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+require 'graphql/client/http'
+
+# The HTTP class
```
Comment (Yegorov) [inline lib/fbe/github/graphql/http.rb:27]: We really need this comment?

File: test/fbe/github/graphql/test_client.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    result = client.query(
+      <<-GRAPHQL
+        query {
+          viewer {
+              login
+          }
+        }
+      GRAPHQL
+    )
+    assert(!result.viewer.login.empty?)
```
Comment (Yegorov) [inline test/fbe/github/graphql/test_client.rb:49]: It is [better to use](https://minitest.rubystyle.guide/#refute-false) `refute(false)`  instead `assert(!true)`.

File: test/fbe/github/graphql/test_client.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+# Author:: Yegor Bugayenko (yegor256@gmail.com)
+# Copyright:: Copyright (c) 2024 Zerocracy
+# License:: MIT
+class TestGitHubGraphQL < Minitest::Test
+  def test_simple_query
+    skip # it's a "live" test, run it manually if you need it
+    WebMock.allow_net_connect!
+    client = Fbe::GitHub::GraphQL::Client.new(token: ENV.fetch('GITHUB_TOKEN', nil))
+    result = client.query(
+      <<-GRAPHQL
```
Comment (Yegorov) [inline test/fbe/github/graphql/test_client.rb:41]: Probably better to use [`<<~`](https://infinum.com/blog/multiline-strings-ruby-2-3-0-the-squiggly-heredoc/)

File: lib/fbe/gh_graphql.rb [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                }
+              }
+            GRAPHQL
+          )
+          result.repository.pull_request.review_threads.to_h['nodes']
+        end
+
+        def total_commits(owner, name, branch)
+          result = @origin.query(
+            <<-GRAPHQL
```
Comment (Yegorov) [inline lib/fbe/gh_graphql.rb:80]: Probably better to use [`<<~`](https://infinum.com/blog/multiline-strings-ruby-2-3-0-the-squiggly-heredoc/)

File: lib/fbe/gh_graphql.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      if options.testing.nil?
+        g = Fbe::GitHub::GraphQL::Client.new(token: options.github_token || ENV.fetch('GITHUB_TOKEN', nil))
+      else
+        loog.debug('The connection to GitHub GraphQL API is mocked')
+        g = Fbe::FakeGitHubGraphQLClient.new
+      end
+      decoor(g, loog:) do
+        def resolved_conversations(owner, name, number)
+          result = @origin.query(
+            <<-GRAPHQL
```
Comment (Yegorov) [inline lib/fbe/gh_graphql.rb:50]: Probably better to use [`<<~`](https://infinum.com/blog/multiline-strings-ruby-2-3-0-the-squiggly-heredoc/)

File: lib/fbe/github/graphql.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# The module.
```
Comment (Yegorov) [inline lib/fbe/github/graphql.rb:25]: We really need this comment and below?

File: lib/fbe/github.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# The module.
```
Comment (Yegorov) [inline lib/fbe/github.rb:25]: We really need this comment and below?

**Review by Yegorov — CHANGES_REQUESTED** [review]
@Suban05 There are several comments
@yegor256 What do you think?

File: lib/fbe/github/graphql.rb (hunk 8 above)
Comment (Suban05) [reply lib/fbe/github/graphql.rb:25]: @Yegorov Rubocop swears if there is no description of the class or module

Comment (Suban05) [reply lib/fbe/github/graphql.rb:25]: @Yegorov below - do you mean defining the `Fbe::GitHub::GraphQL` module?

**Review by Yegorov — APPROVED** [review]
@Suban05 Looks fine!

**Comment by Suban05** [comment]
@yegor256 please check

**Comment by yegor256** [comment]
@Suban05 looks very good! However, for such a small project, do we really need such a deep hierarchy of packages? Maybe you can keep these supplementary classes inside the `Fbe` package?

**Comment by Suban05** [comment]
@yegor256 do I understand correctly that you mean to simplify such a hierarchy `Fbe::GitHub::GraphQL::Client` to the next `Fbe::GitHubGraphQLClient`? 
I'm not sure about the name

**Comment by yegor256** [comment]
@Suban05 yes, exactly. Usually, we need packages when a decent number of classes may be moved there (at least, I would say, 10).

**Comment by Suban05** [comment]
@yegor256 fixed

**Comment by yegor256** [comment]
@Suban05 looks much better, but I still don't like the naming and allocation of files. In this `Fbe` gem we have one file per one "thing", just like `octo.rb` and others. In this pull request, however, you introduce one thing (GitHub GraphQL), but there are three files for it. Would be more in line with our overall design to put all classes into one file, I believe. Also, I would suggest making it a class:

```
def github_graph(options:)
  global[:github_graph] ||= begin
    if options.testing
      Fbe::Graph.new(token, host, port)
    else
      Fbe::Graph::Fake.new
    end
  end
end
class Fbe::Graph
  def resolved_conversations
  def total_commits
  class Fake
end
```

I believe, `github_graph` is a better name than `gh_graphql`, because the user doesn't care about the QL part, but only about the essence of the entity. It is a graph. Not a query language. How exactly we communicate with the graph, it is our internal details. For the user it is a **graph**.

**Comment by Suban05** [comment]
@yegor256 fixed, please check

**Comment by yegor256** [comment]
@Suban05 now it looks perfect. Hope, you also like it :) Thanks!

**Comment by 0crat** [comment]
@Yegorov 🎉 Great job on the review! You've earned a whopping **38 points**: 15 base points, 7 for reviewing 382 lines of code, and 16 for your insightful 16 comments. Your dedication is paying off – your balance now stands at [**+329 points**](https://zerocracy.github.io/judges-action/zerocracy-vitals.html). Keep up the fantastic work!

**Comment by 0crat** [comment]
@Suban05 Hey there, great work on your contribution! 👍 You've snagged +7 points for this one. Here's the breakdown: +20 base points, -7 for the 230 [hits-of-code](https://www.yegor256.com/2014/11/14/hits-of-code.html) (exceeding the 200 limit), and -5 for the 21 review comments. Remember, [keeping it concise](https://www.yegor256.com/2018/03/06/speed-vs-quality.html) can boost your points next time. Keep the code coming – your balance is now at [+411](https://zerocracy.github.io/judges-action/zerocracy-vitals.html)!

**Pull request merged**